Clarity standard type coercion for function calls with multiple outputs (#3888)
modelica/ModelicaSpecification: chapters/interface.tex

    If both \lstinline!size(A)! and \lstinline!size(B)! are known and \lstinline!size(A)! = \lstinline!size(B)! then this defines the size of the full expression, otherwise the size of the full expression is not known until the expression is about to be evaluated.
    In case of an \lstinline!if!-expression the size of the full expression is defined based on the branch selected, and for other cases \lstinline!size(A)! = \lstinline!size(B)! must hold at this point.
  \item
    If \lstinline!A! is a scalar expression of a simple type \lstinline!B! must also be a scalar expression of a simple type.
  \end{itemize}
\item
  If \lstinline!A! is a \lstinline!Real! expression then \lstinline!B! must be a \lstinline!Real! or \lstinline!Integer! expression.
  The type of the full expression is \lstinline!Real!, compare \cref{standard-type-coercion}, unless the operator is a relational operator (\cref{equality-relational-and-logical-operators}) where the type of the full expression is \lstinline!Boolean!.
\item
  If \lstinline!A! is an \lstinline!Integer! expression:
  \begin{itemize}
  \item

>     In equations receiving several outputs of a function call and assignment statements, and where \lstinline!A! is part of the left-hand-side and \lstinline!B! part of the right-hand-side, \lstinline!B! must be an \lstinline!Integer! expression.

Reviewer: The current text still doesn't say how A and B are related, other than being on different sides of the equation/assignment. 8.3.1 and 11.2.1.1 specify how the lhs and rhs are related though, so just adding "the corresponding" here should be enough to fix that.
```suggestion
    In equations receiving several outputs of a function call and assignment statements, and where \lstinline!A! is part of the left-hand-side and \lstinline!B! the corresponding part of the right-hand-side, \lstinline!B! must be an \lstinline!Integer! expression.
```

>     For assignment statements and equations receiving several outputs of a function call, and where \lstinline!A! is part of the left-hand-side and \lstinline!B! part of the right-hand-side, \lstinline!B! must be an \lstinline!Integer! expression.

Reviewer 2: ```suggestion
    For assignment statements and equations receiving several outputs of a function call, and where \lstinline!A! is part of the left-hand-side and \lstinline!B! part of the right-hand-side, \lstinline!B! must be an \lstinline!Integer! expression, and the type of the full expression is \lstinline!Integer!.
```

Author: I omitted this intentionally, since I didn't think it made sense to speak of the type of the assignment or equation as a whole, as these are not expressions?

Reviewer 2: Well, the type does matter for the type-split "Local equation size", even though it is already defined as using the lhs-variable type.

Reviewer: I don't see any issue with @HansOlsson's suggestion, since it's talking about the resulting type of A and B and not the whole assignment or equation.

Reviewer: Maybe it would be better to say "In assignment statements and equation", rather than "For ..."?

Reviewer 2: Possible, and could also swap from:

> In assignment statements and equations receiving several outputs of a function call,

to:

> In equations receiving several outputs of a function call and assignment statements,

As it relates to all assignment statements, not only the ones with several outputs.

Author: > In equations receiving several outputs of a function call and assignment statements,

This is now done.

Author: > I don't see any issue with @HansOlsson's suggestion, since it's talking about the resulting type of A and B and not the whole assignment or equation.

Is it really?  I thought it was talking about the type of "whole expressions" such as `A + B`?  That is, that the type of `1 + 1.5` is `Real`.

Reviewer: > > I don't see any issue with @HansOlsson's suggestion, since it's talking about the resulting type of A and B and not the whole assignment or equation.
> 
> Is it really? I thought it was talking about the type of "whole expressions" such as `A + B`? That is, that the type of `1 + 1.5` is `Real`.

I was thinking about it in more in terms of `(A1, A2, ...) := (B1, B2, ...)`, and what the "compatible type" of e.g. A1 and B1 would be. But if we forbid type coercion in this case, then I guess it doesn't really matter anymore.

>     For assignment statements and equations receiving several outputs of a function call, and where \lstinline!A! is part of the left-hand-side and \lstinline!B! part of the right-hand-side, \lstinline!B! must be an \lstinline!Integer! expression.

Reviewer: This doesn't say how the expressions on the lhs and rhs are related to each other, and "part of" is a bit ambiguous. Maybe something like this instead?
```suggestion
    For assignment statements and equations receiving several outputs of a function call, and where \lstinline!A! is one of the component references (or derivatives) on the left-hand-side and \lstinline!B! the corresponding function call result on the right-hand-side, \lstinline!B! must be an \lstinline!Integer! expression.
```

Reviewer 2: I don't think this change is needed. 

If needed it would require more changes since
* **The text has been updated.**
* It should be one output from the corresponding function call on the right-hand-side, not just the function call.
* "Part of" kind of also work for record members in the left-hand-side; even if they are not the component-reference on the lhs.

Reviewer 2: Can you confirm @perost ?
